BJS2-77986. Подписка на других пользователей

Keriod opened the pull request:
Реализованы котроллер и сервис для подписки/отписки, получения списка подписчиков и их кол-ва

Keriod committed:
BJS2-77986. Подписаться на пользователя

Keriod committed:
BJS2-77986. Отписаться от пользователя

Keriod committed:
BJS2-77986. Подписка на других пользователей

Keriod committed:
BJS2-77986. Подписка на других пользователей

psvm-dev commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,68 @@
+
+    @PostMapping("/{followerId}/follow/{followeeId}")
+    public void followUser(@PathVariable("followerId") long followerId, @PathVariable("followeeId") long followeeId) {
+        if (checkFollowForSelf(followerId, followeeId)) {
```
надо бы вынести логику валидации из контроллера и в самом контроллере валидировать только при помощи аннотации @Valid

psvm-dev commented on src/main/java/school/faang/user_service/dto/UserDto.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
```
кажется если в поставить NoArgsConstructor, то ReqArgsConst из Data "перетрётся". не помню, точно ли это так. но если так, то надо бы явно прописать @ReqArgsConst

Keriod committed:
BJS2-77986. Подписка на других пользователей

1) Написание класса-валидатора для подписок
2) Проверка в фильтрах на null

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,55 @@
+
+    private final SubscriptionService subscriptionService;
+
+    @PostMapping("/{followerId}/follow/{followeeId}")
+    public void followUser(@PathVariable("followerId") long followerId, @PathVariable("followeeId") long followeeId) {
+        subscriptionService.followUser(followerId, followeeId);
+    }
```
у тебя контроллер подписок, но подписки не существуют без привязки к пользователю, значит подписки - это подпуть юзера. 
followerId / followeeId - это же юзеры?

я бы делал 
/api/v1/users/{userId}/subscriptions - список моих подписок
/api/v1/users/{userId}/subscriptions/follow?followerId={id} - подписаться на кого-то

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @DeleteMapping("/{followerId}/unfollow/{followeeId}")
+    public void unfollowUser(@PathVariable("followerId") long followerId, @PathVariable("followeeId") long followeeId) {
```
не надо void, возвращай ResponseEntity.ok

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @GetMapping("{followeeId}/followers")
+    public List<UserDto> getFollowers(@PathVariable("followeeId") long followeeId,
```
Везде ответы оборачивай в ResponseEntity

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,55 @@
+@RestController
+@RequestMapping("/subscriptions")
+@RequiredArgsConstructor
+public class SubscriptionController {
```
путь урлов переработай

20cognosce commented on src/main/java/school/faang/user_service/mappers/user/UserMapper.java:
```diff
@@ -0,0 +1,12 @@
+
+@Mapper(componentModel = "spring")
+public interface UserMapper {
+    UserDto userToUserDTO(User user);
+
+    User userDTOToUser(UserDto dto);
```
И что, и это всё? Ты проверил, что все поля мапятся как надо?

20cognosce commented on src/main/java/school/faang/user_service/dto/user/UserDto.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+@NoArgsConstructor
+public class UserDto {
+    @NotNull
+    private Long id;
+    @NotBlank
+    private String username;
+    @NotBlank
+    private String email;
```
Это не полная инфа о пользователе

20cognosce replied:
У тебя будут конфликты с другими DTOшками, если это чисто под подписки, то и называй SubscriptionDto

20cognosce commented on src/main/java/school/faang/user_service/service/subscriptions/SubscriptionService.java:
```diff
@@ -0,0 +1,88 @@
+public class SubscriptionService {
+
+    private final SubscriptionRepository subscriptionRepository;
+    private final UserMapper userMapper;
```
сервис не должен иметь зависимость на маппер, мапь в контроллере

20cognosce commented on src/main/java/school/faang/user_service/validation/subscriptions/SubscriptionValidations.java:
```diff
@@ -0,0 +1,36 @@
+import school.faang.user_service.exceptions.DataValidationException;
+
+@Component
+public class SubscriptionValidations {
```
Validator, это компонент, единственное число

20cognosce commented on src/main/java/school/faang/user_service/validation/subscriptions/SubscriptionValidations.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    public void ValidateUnfollowAction(long followerId, long followeeId) {
```
почему у тебя методы с большой буквы?

20cognosce commented on src/main/java/school/faang/user_service/validation/subscriptions/SubscriptionValidations.java:
```diff
@@ -0,0 +1,36 @@
+import org.springframework.stereotype.Component;
+import school.faang.user_service.exceptions.DataValidationException;
+
+@Component
```
Тебе бин тут фактически не нужен, делай статические методы

20cognosce commented on src/main/java/school/faang/user_service/service/subscriptions/SubscriptionService.java:
```diff
@@ -0,0 +1,88 @@
+        subscriptionValidations.ValidateFollowAction(followerId, followeeId);
+        subscriptionValidations.ValidateSubscribeAction(existSub);
+
+        subscriptionRepository.followUser(followerId, followeeId);
```
Что там в репозитории происходит, ты проверил создание записи в БД? Через Postman протестил все методы API?
Сущности Subscription разве нет никакой?

Keriod replied:
Проверял все методы через Postman и видел результаты отработки в БД, все было нормально. В пакете entity нет сущности Subscription

20cognosce replied:
Мне интересно, что там в запросе происходит, какой insert
Что такое вообще подписка на пользователя в схеме БД

Keriod replied:
В БД есть таблица subscription с полями follower_id, followee_id, created_at, updated_at. Получается связь id юзеров и когда она была создана. В insert залетают только follower_id и followee_id, а даты автоматом записываются как текущее время

20cognosce replied:
Ну то есть запись в связной таблице обозначает подписку
Можно сущность выделить под это и работать через save / delete, а не нативками

20cognosce commented on src/main/java/school/faang/user_service/service/subscriptions/SubscriptionService.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Transactional(readOnly = true)
+    public Integer getFollowersCount(long followeeId) {
```
Long

20cognosce reviewed (changes requested):
в целом похоже на правду, доделывай и пойдём дальше

psvm-dev commented on src/main/java/school/faang/user_service/service/subscriptions/SubscriptionService.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Transactional(readOnly = true)
+    public List<UserDto> getFollowing(long followeeId, UserFilterDto filterDto) {
+        Stream<User> userStream = subscriptionRepository.findByFollowerId(followeeId);
```
нужно сделать обработку на null. если записи нет, то вернуть NotFoundException

Keriod committed:
BJS2-77986. Подписка на других пользователей

1) Изменены url в контроллере
2) Маппинг теперь в котроллере
3) Переписан маппер

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,92 @@
+    public ResponseEntity<String> followUser(@PathVariable("userId") long followerId, @RequestParam long followeeId) {
+        subscriptionService.followUser(followerId, followeeId);
+        return new ResponseEntity<>(
+                "You followed user with id = " + followeeId,
```
не надо такое на фронт возвращать, как это потом парсить? только строгие данные, без общения с пользователем

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,92 @@
+        );
+    }
+
+    @GetMapping("/{userId}/subscriptions/followers")
+    public ResponseEntity<List<SubscriptionDto>> getFollowers(@PathVariable("userId") long followeeId,
+                                                              @ModelAttribute SubscriptionFilterDto filterDto) {
```
Я не понял, причем тут @ModelAttribute, ведь это относится к формам в HTML, как ты через постман эту DTO передаешь, в теле запроса?
Если в теле, то GET не подходит, делай POST

20cognosce commented on src/main/java/school/faang/user_service/dto/subscription/SubscriptionDto.java:
```diff
@@ -0,0 +1,19 @@
+@NoArgsConstructor
+public class SubscriptionDto {
+    @NotNull
+    private Long id;
```
userId

20cognosce commented on src/main/java/school/faang/user_service/mappers/subscription/SubscriptionMapper.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public User SubscriptionDtoToUser(SubscriptionDto dto) {
+        User user = new User();
+        user.setId(dto.getId());
+        user.setUsername(dto.getUsername());
+        user.setEmail(dto.getEmail());
```
@Builder

20cognosce commented on src/main/java/school/faang/user_service/service/subscriptions/SubscriptionService.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Transactional(readOnly = true)
+    public Long getFollowersCount(long followeeId) {
+        return (long) subscriptionRepository.findFollowersAmountByFolloweeId(followeeId);
```
не надо кастовать, репозиторий измени

Keriod committed:
BJS2-77986. Подписка на других пользователей

1) PR правки

20cognosce commented on src/main/java/school/faang/user_service/mappers/subscription/SubscriptionMapper.java:
```diff
@@ -0,0 +1,26 @@
+import school.faang.user_service.dto.subscription.SubscriptionDto;
+import school.faang.user_service.entity.User;
+
+@Component
+public class SubscriptionMapper {
```
можно статикой

20cognosce commented on src/main/java/school/faang/user_service/controller/subscriptions/SubscriptionController.java:
```diff
@@ -0,0 +1,91 @@
+public class SubscriptionController {
+
+    private final SubscriptionService subscriptionService;
+    private final SubscriptionMapper subscriptionMapper;
```
Если маппер статический, то зависимости не понадобятся
